codec.go

```diff
@@ -1,40 +1,100 @@
 package rpc
 
 import (
-	"github.com/ugorji/go/codec"
 	"io"
+
+	"github.com/ugorji/go/codec"
 )
 
 type decoder interface {
 	Decode(interface{}) error
 }
 
-type byteReadingDecoder interface {
-	decoder
-	io.ByteReader
+type encoder interface {
+	Encode(interface{}) <-chan error
```

Comment:
I've added a TODO, but `decoder` is actually just that, whereas `encoder` is the one doing multiple things.